core/src/main/java/com/google/errorprone/bugpatterns/AssertFalse.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.errorprone.bugpatterns;
+
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker.AssertTreeMatcher;
+import com.google.errorprone.fixes.Fix;
+import com.google.errorprone.fixes.SuggestedFix;
+import com.google.errorprone.matchers.Description;
+
+import com.sun.source.tree.AssertTree;
+import com.sun.source.tree.ExpressionTree;
+
+import static com.google.errorprone.BugPattern.Category.JDK;
+import static com.google.errorprone.BugPattern.MaturityLevel.EXPERIMENTAL;
+import static com.google.errorprone.BugPattern.SeverityLevel.WARNING;
+import static com.google.errorprone.matchers.Matchers.kindIs;
+import static com.sun.source.tree.Tree.Kind.BOOLEAN_LITERAL;
+import static com.sun.tools.javac.tree.JCTree.JCLiteral;
+
+/**
+ * @author sebastian.h.monte@gmail.com (Sebastian Monte)
+ */
+@BugPattern(name = "AssertFalse",
+    summary = "Assert false should not be used",
+    explanation = "Assert false indicates that the code should never be"
+                  + " executed except in case of a bug. It is better to"
+                  + " throw an AssertionError so an exception is raised"
+                  + " regardless whether assertions are enabled.",
+    category = JDK, severity = WARNING, maturity = EXPERIMENTAL)
+public class AssertFalse extends BugChecker implements AssertTreeMatcher {
+
+  @Override
+  public Description matchAssert(AssertTree tree, VisitorState state) {
+    ExpressionTree condition = tree.getCondition();
+    if (kindIs(BOOLEAN_LITERAL).matches(condition, state)) {
```

Comment:
You can replace this with `Matchers.booleanLiteral(false)`.  Also, I think it would be generally useful to have a matcher that takes a AssertTree and a Matcher\<ExpressionTree\>, and applies the matcher to the condition.  Could you add something like that to the Matchers class, and use it here?  It would look something like:
```java
public static Matcher<AssertTree> assertionWithCondition(final Matcher<ExpressionTree> conditionMatcher) {
  return new Matcher<AssertTree>() {
    @Override
    public boolean matches(AssertTree tree, VisitorState state) {
      return conditionMatcher.matches(tree.getCondition(), state);
    }
  };
}
```